File: 24_高级数据结构/good java/BitCountProblem/BitCountProblem.java
package template.problem;

import template.binary.Log2;

public class BitCountProblem {
    public static final long INF = (long) 2e18;

    public static long countZero(long n) {
        return countLength(n) - countOne(n);
    }

    public static long countLength(long n) {
        if (n == 0) {
            return 1;
        }
        long ans = 0;
        int len = Log2.floorLog(n);
        for (int i = 0; i <= len; i++) {
            long l = 1L << i;
            long r = Math.min(l * 2 - 1, n);
            ans += (i + 1) * (r - l + 1);
        }
        return ans;
    }

    /**
     * count how many 1 occur in binary form of 0, 1, ..., n
     */
    public static long countOne(long n) {
        if (n == 0) {
            return 0;
        }
        long ans = 0;
        int len = Log2.floorLog(n);
        for (int i = 0; i <= len; i++) {
            long bit = 1L << i;
            long remain = n;
            long block = remain / (2 * bit);
            ans += block * bit;
            remain %= (2 * bit);
            ans += Math.max(0, remain - bit + 1);
        }
        return ans;
    }

    public static long countOne(long l, long r) {
        long ans = countOne(r);
        if (l > 0) {
            ans -= countOne(l - 1);
        }
        return ans;
    }

    public static long countLength(long l, long r) {
        long ans = countLength(r);
        if (l > 0) {
            ans -= countLength(l - 1);
        }
        return ans;
    }

    public static long countZero(long l, long r) {
        long ans = countZero(r);
        if (l > 0) {
            ans -= countZero(l - 1);
        }
        return ans;
    }

}

File: 24_高级数据结构/good java/BitCountProblem/BitCountProblem.go
// BitCountProblem

package main

import (
	"fmt"
	"math/bits"
)

func main() {
	fmt.Println(CountZeros(10))
	fmt.Println(CountOnes(10))
	fmt.Println(CountLength(10))

	fmt.Println(CountZerosRange(3, 10))
	fmt.Println(CountOnesRange(3, 10))
	fmt.Println(CountLengthRange(3, 10))
}

// [0, n] 之间的数中二进制表示中0的个数之和.
func CountZeros(n int) int {
	return CountLength(n) - CountOnes(n)
}

// [0, n] 之间的数中二进制表示中1的个数之和.
func CountOnes(n int) int {
	if n == 0 {
		return 0
	}
	res := 0
	len := floorLog64(uint64(n))
	for i := 0; i <= len; i++ {
		bit := 1 << i
		remain := n
		block := remain / (2 * bit)
		res += block * bit
		remain %= (2 * bit)
		res += max(0, remain-bit+1)
	}
	return res
}

// [0, n] 之间的数中二进制表示位数之和.
func CountLength(n int) int {
	if n == 0 {
		return 1
	}
	res := 0
	len := floorLog64(uint64(n))
	for i := 0; i <= len; i++ {
		l := 1 << i
		r := min(l*2-1, n)
		res += (i + 1) * (r - l + 1)
	}
	return res
}

// [left, right] 之间的数中二进制表示中0的个数之和.
func CountZerosRange(left, right int) int {
	res := CountZeros(right)
	if left > 0 {
		res -= CountZeros(left - 1)
	}
	return res
}

// [left, right] 之间的数中二进制表示中1的个数之和.
func CountOnesRange(left, right int) int {
	res := CountOnes(right)
	if left > 0 {
		res -= CountOnes(left - 1)
	}
	return res
}

// [left, right] 之间的数中二进制表示位数之和.
func CountLengthRange(left, right int) int {
	res := CountLength(right)
	if left > 0 {
		res -= CountLength(left - 1)
	}
	return res
}

func floorLog64(x uint64) int {
	if x <= 0 {
		panic("IllegalArgumentException")
	}
	return 63 - bits.LeadingZeros64(x)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
